src/Rule/CompareHandler.php

```diff
@@ -77,23 +80,68 @@ public function validate(mixed $value, object $rule, ValidationContext $context)
      */
     private function compareValues(string $operator, string $type, mixed $value, mixed $targetValue): bool
     {
-        if ($type === CompareType::NUMBER) {
-            $value = (float) $value;
-            $targetValue = (float) $targetValue;
-        } else {
-            $value = (string) $value;
-            $targetValue = (string) $targetValue;
+        if (!in_array($operator, ['==', '===', '!=', '!=='])) {
+            [$value, $targetValue] = $this->getTypeCastedValues($type, $value, $targetValue);
         }
 
         return match ($operator) {
-            '==' => $value == $targetValue,
-            '===' => $value === $targetValue,
-            '!=' => $value != $targetValue,
-            '!==' => $value !== $targetValue,
+            '==' => $this->assertEquals($type, $value, $targetValue),
+            '===' => $this->assertEquals($type, $value, $targetValue, strict: true),
+            '!=' => $this->assertNotEquals($type, $value, $targetValue),
+            '!==' => $this->assertNotEquals($type, $value, $targetValue, strict: true),
             '>' => $value > $targetValue,
             '>=' => $value >= $targetValue,
             '<' => $value < $targetValue,
             '<=' => $value <= $targetValue,
         };
     }
+
+    private function assertEquals(
+        string $type,
+        mixed $value,
+        mixed $targetValue,
+        bool $strict = false,
+    ): bool {
+        if ($strict && gettype($value) !== gettype($targetValue)) {
+            return false;
+        }
+
+        if ($type === CompareType::NUMBER) {
+            return $this->assertFloatsEqual((float) $value, (float) $targetValue);
+        }
+
+        return (string) $value == (string) $targetValue;
```

Comment:
Are you sure all types should be compared like that except numbers?

Reply:
what is the reason to cast to string?

Reply:
That how it was in previous version. Do you have edge case examples?

Reply:
> That how it was in previous version. Do you have edge case examples?

For example, compare objects with `===`

Reply:
@vjik Objects are not supported for both rule value and input value:

https://github.com/yiisoft/validator/blob/7be3488331d9348cb3e7e247f73ddac8214d5f14/src/Rule/AbstractCompare.php#L111-L112

https://github.com/yiisoft/validator/blob/7be3488331d9348cb3e7e247f73ddac8214d5f14/src/Rule/AbstractCompare.php#L111-L112